libraries/AP_HAL_PX4/CAN.h

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright (C) 2014 Pavel Kirienko <pavel.kirienko@gmail.com>
+ *
+ * With modifications for Ardupilot CAN driver
+ * Copyright (C) 2017 Eugene Shamaev
+ */
+
+#pragma once
+
+#include "AP_HAL_PX4.h"
+#include <systemlib/perf_counter.h>
+#include <AP_HAL/CAN.h>
+#include <pthread.h>
+#include <semaphore.h>
+
+#include "bxcan.h"
+#include "AP_HAL/utility/RingBuffer.h"
+
+#define UAVCAN_STM32_NUM_IFACES 2
+#define UAVCAN_STM32_RX_QUEUE_SIZE 64
+
+#ifndef UAVCAN_NULLPTR
+# if UAVCAN_CPP_VERSION >= UAVCAN_CPP11
+#  define UAVCAN_NULLPTR nullptr
+# else
+#  define UAVCAN_NULLPTR NULL
+# endif
+#endif
+
+/**
+ * IRQ handler macros
+ */
+# define UAVCAN_STM32_IRQ_HANDLER(id)  int id(int irq, FAR void* context)
+# define UAVCAN_STM32_IRQ_PROLOGUE()
+# define UAVCAN_STM32_IRQ_EPILOGUE()    return 0;
+
+/**
+ * Glue macros
+ */
+#define UAVCAN_STM32_GLUE2_(A, B)       A##B
+#define UAVCAN_STM32_GLUE2(A, B)        UAVCAN_STM32_GLUE2_(A, B)
+
+#define UAVCAN_STM32_GLUE3_(A, B, C)    A##B##C
+#define UAVCAN_STM32_GLUE3(A, B, C)     UAVCAN_STM32_GLUE3_(A, B, C)
+
+namespace PX4 {
+/**
+ * Driver error codes.
+ * These values can be returned from driver functions negated.
+ */
+static const int16_t ErrUnknown = 1000; ///< Reserved for future use
+static const int16_t ErrNotImplemented = 1001; ///< Feature not implemented
+static const int16_t ErrInvalidBitRate = 1002; ///< Bit rate not supported
+static const int16_t ErrLogic = 1003; ///< Internal logic error
+static const int16_t ErrUnsupportedFrame = 1004; ///< Frame not supported (e.g. RTR, CAN FD, etc)
+static const int16_t ErrMsrInakNotSet = 1005; ///< INAK bit of the MSR register is not 1
+static const int16_t ErrMsrInakNotCleared = 1006; ///< INAK bit of the MSR register is not 0
+static const int16_t ErrBitRateNotDetected = 1007; ///< Auto bit rate detection could not be finished
+
+/**
+ * RX queue item.
+ * The application shall not use this directly.
+ */
+struct CanRxItem {
+    uint64_t utc_usec;
+    uavcan::CanFrame frame;
+    uavcan::CanIOFlags flags;
+    CanRxItem() :
+        utc_usec(0), flags(0)
+    {
+    }
+};
+
+struct CriticalSectionLocker {
+    const irqstate_t flags_;
+
+    CriticalSectionLocker() :
+        flags_(irqsave())
+    {
+    }
+
+    ~CriticalSectionLocker()
+    {
+        irqrestore(flags_);
+    }
+};
+
+namespace clock {
+uint64_t getUtcUSecFromCanInterrupt();
```

Comment:
This is connected with "FOR INVESTIGATION" comment.
I would prefer to leave it as is for now.